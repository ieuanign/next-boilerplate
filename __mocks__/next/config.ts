const { publicRuntimeConfig } = require("../../next.config");

const getConfig = () => ({
	publicRuntimeConfig,
	serverRuntimeConfig: publicRuntimeConfig,
});

export default getConfig;
