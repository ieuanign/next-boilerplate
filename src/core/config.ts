import getConfig from "next/config";

interface PublicRuntimeConfig {
	environment: string;
	apiHost: string;
	mockAPI: boolean;
}

const {
	publicRuntimeConfig: env,
}: { publicRuntimeConfig: PublicRuntimeConfig } = getConfig();

const apiBase = (pathname: string) => {
	return new URL(pathname, env.apiHost).toString();
};

const staticPath = (filename: string) => {
	if (filename.startsWith("/")) {
		filename = filename.substring(1);
	}

	return `/static/${filename}`;
};

export { env, apiBase, staticPath };
