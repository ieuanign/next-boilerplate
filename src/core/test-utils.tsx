import { GetServerSidePropsContext } from "next";
import { Router } from "next/router";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import App from "@pages/_app.page";

const useRouter = jest.spyOn(require("next/router"), "useRouter");

jest.mock("react-i18next", () => ({
	useTranslation: () => ({ t: (key: any) => key }),
}));

export const setup = (Component: Function, defaultRouter?: Partial<Router>) => {
	let mockRouter = {
		pathname: "/",
		query: {},
		...defaultRouter,
		push: jest.fn(),
		prefetch: jest.fn(() => Promise.resolve(true)),
	};

	const withProps = (props?: any) => {
		const {
			hostname = "http://localhost:3000",
			pageProps = {},
			router,
			...rest
		} = props || {};

		mockRouter = { ...mockRouter, ...router };

		useRouter.mockReturnValue(mockRouter);

		const Comp = render(
			<App
				pageProps={pageProps}
				hostname={hostname}
				router={mockRouter}
				{...rest}
				Component={Component}
			/>
		);

		return {
			router: mockRouter,
			userEvent: userEvent.setup(),
			...Comp,
		};
	};

	return withProps;
};

export const getContext = (
	pathname: string,
	config?: any
): GetServerSidePropsContext =>
	({
		query: {},
		locale: "en",
		...config,
		resolvedUrl: pathname,
		req: {
			...config?.req,
			cookies: {
				...config?.req?.cookies,
			},
		},
	} as GetServerSidePropsContext);
