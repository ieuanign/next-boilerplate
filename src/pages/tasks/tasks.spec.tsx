import { screen, waitFor } from "@testing-library/react";
import { rest } from "msw";

import Tasks, { getServerSideProps } from "./index.page";

import { server } from "@mocks/server";
import StatusCodes from "@core/http/status-codes";
import { apiBase } from "@core/config";

import { init, getContext } from "@core/test-utils";

const setup = init(Tasks, {
	pathname: "/en/tasks",
});

describe("getServerSideProps", () => {
	const defaultContext = getContext("/en/tasks");

	it("with fallback data", async () => {
		const ssrProps = await getServerSideProps(defaultContext);

		// show fallback data
		expect(ssrProps).toMatchSnapshot();
	});

	it("without fallback data", async () => {
		server.use(
			rest.get(apiBase("/tasks"), async (_, res, ctx) =>
				res(ctx.status(StatusCodes.INTERNAL_SERVER_ERROR))
			)
		);

		const ssrProps = await getServerSideProps(defaultContext);

		// empty fallback
		expect(ssrProps).toMatchSnapshot();
	});

	// for default ssrLangProps utility
	it("without locale", async () => {
		const customContext = {
			...defaultContext,
			locale: undefined,
		};

		const ssrProps = await getServerSideProps(customContext);

		// show fallback data
		expect(ssrProps).toMatchSnapshot();
	});
});

const fallback = {
	'"@"http://localhost:3080/tasks","': {
		data: [
			{
				id: "63b8bb76-cf6a-4972-a65a-b5730b69ba27",
				title: "quidem expedita labore",
			},
			{
				id: "b86a4d7b-b95b-465a-9533-4fac1c606272",
				title: "dolores veritatis voluptatum",
			},
			{
				id: "e6cebf88-7991-4457-aade-4177f3869c17",
				title: "vel inventore quaerat",
			},
			{
				id: "48650456-bade-4487-a49d-94a8d215ce32",
				title: "assumenda quod unde",
			},
			{
				id: "56034668-b8f0-45ac-890b-e246c355d80a",
				title: "fugiat quibusdam expedita",
			},
			{
				id: "2e57a3d4-8dd8-46a5-85d2-6d0578899840",
				title: "quos beatae adipisci",
			},
			{
				id: "e5384784-e4f9-44b9-8c56-abd229fd5fd1",
				title: "repudiandae doloribus animi",
			},
			{
				id: "3e7e8310-4de3-4907-9c7b-806b4d62ccc4",
				title: "quae vel alias",
			},
		],
		error: null,
		status: "ok",
	},
};

describe("Tasks", () => {
	const globalConsole = {
		...global.console,
	};

	beforeAll(() => {
		global.console = {
			...global.console,
			// disable console.error to print up error in unit test
			error: jest.fn(),
		};
	});

	afterAll(() => {
		global.console = {
			...globalConsole,
		};
	});

	it("renders title", () => {
		setup();

		expect(screen.getByText("To Do List")).toBeInTheDocument();
	});

	it("renders with fallback data", async () => {
		setup({
			pageProps: {
				fallback,
			},
		});

		await waitFor(() => {
			expect(screen.getByText("quidem expedita labore")).toBeInTheDocument();
			expect(screen.getAllByText("delete")).toHaveLength(8);
		});
	});

	it("renders error boundary", async () => {
		server.use(
			rest.get(apiBase("/tasks"), async (_, res, ctx) =>
				res(ctx.status(StatusCodes.INTERNAL_SERVER_ERROR))
			)
		);

		setup();

		await waitFor(() => {
			expect(
				screen.getByText(/Error Caught by ErrorBoundary/)
			).toBeInTheDocument();
		});
	});

	describe("delete", () => {
		it("succeed", async () => {
			const { userEvent } = setup();

			await waitFor(() => {
				expect(screen.getByText("quidem expedita labore")).toBeInTheDocument();
				expect(screen.getAllByText("delete")).toHaveLength(8);
			});

			await userEvent.click(screen.getAllByText("delete")[0]);

			await waitFor(() => {
				expect(
					screen.queryByText("quidem expedita labore")
				).not.toBeInTheDocument();
				expect(screen.getAllByText("delete")).toHaveLength(7);
			});
		});

		it("failed", async () => {
			server.use(
				rest.delete(apiBase("/tasks/:id"), async (_, res, ctx) =>
					res(ctx.status(StatusCodes.INTERNAL_SERVER_ERROR))
				)
			);

			const { userEvent } = setup();

			await waitFor(() => {
				expect(screen.getAllByText("delete")).toHaveLength(8);
				expect(screen.getByText("quidem expedita labore")).toBeInTheDocument();
			});

			await userEvent.click(screen.getAllByText("delete")[0]);

			await waitFor(() => {
				expect(screen.getByText("quidem expedita labore")).toBeInTheDocument();
				expect(screen.getAllByText("delete")).toHaveLength(8);
				expect(console.error).toHaveBeenCalled();
			});
		});
	});

	describe("add", () => {
		it("succeed", async () => {
			const { userEvent } = setup();

			userEvent.type(
				screen.getByPlaceholderText("Enter task name"),
				"Swimming"
			);

			await waitFor(() => {
				expect(screen.getByPlaceholderText("Enter task name")).toHaveValue(
					"Swimming"
				);
			});

			await userEvent.click(screen.getByText("Add Task"));

			await waitFor(() => {
				expect(screen.getByText("Swimming")).toBeInTheDocument();
				expect(screen.getAllByText("delete")).toHaveLength(9);
				expect(screen.getByPlaceholderText("Enter task name")).toHaveValue("");
			});
		});

		it("failed", async () => {
			server.use(
				rest.post(apiBase("/tasks"), async (_, res, ctx) =>
					res(ctx.status(StatusCodes.INTERNAL_SERVER_ERROR))
				)
			);

			const { userEvent } = setup();

			userEvent.type(
				screen.getByPlaceholderText("Enter task name"),
				"Swimming"
			);

			await waitFor(() => {
				expect(screen.getByPlaceholderText("Enter task name")).toHaveValue(
					"Swimming"
				);
			});

			await userEvent.click(screen.getByText("Add Task"));

			await waitFor(() => {
				expect(screen.queryByText("Swimming")).not.toBeInTheDocument();
				expect(screen.getAllByText("delete")).toHaveLength(8);
				expect(screen.getByPlaceholderText("Enter task name")).toHaveValue(
					"Swimming"
				);
			});
		});
	});
});
