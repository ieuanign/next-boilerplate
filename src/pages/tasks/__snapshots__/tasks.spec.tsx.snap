// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getServerSideProps with fallback data 1`] = `
{
  "props": {
    "_nextI18Next": {
      "initialI18nStore": {
        "en": {
          "common": {
            "description": "NextJS common setup",
            "no": "No",
            "title": "NextJS Boilerplate | EN",
            "yes": "Yes",
          },
        },
      },
      "initialLocale": "en",
      "ns": [
        "common",
      ],
      "userConfig": {
        "default": {
          "i18n": {
            "defaultLocale": "default",
            "localeDetection": false,
            "locales": [
              "default",
              "en",
              "id",
            ],
          },
        },
        "i18n": {
          "defaultLocale": "default",
          "localeDetection": false,
          "locales": [
            "default",
            "en",
            "id",
          ],
        },
      },
    },
    "fallback": {
      "@"http://localhost:3080/tasks",": {
        "data": [
          {
            "id": "63b8bb76-cf6a-4972-a65a-b5730b69ba27",
            "title": "quidem expedita labore",
          },
          {
            "id": "b86a4d7b-b95b-465a-9533-4fac1c606272",
            "title": "dolores veritatis voluptatum",
          },
          {
            "id": "e6cebf88-7991-4457-aade-4177f3869c17",
            "title": "vel inventore quaerat",
          },
          {
            "id": "48650456-bade-4487-a49d-94a8d215ce32",
            "title": "assumenda quod unde",
          },
          {
            "id": "56034668-b8f0-45ac-890b-e246c355d80a",
            "title": "fugiat quibusdam expedita",
          },
          {
            "id": "2e57a3d4-8dd8-46a5-85d2-6d0578899840",
            "title": "quos beatae adipisci",
          },
          {
            "id": "e5384784-e4f9-44b9-8c56-abd229fd5fd1",
            "title": "repudiandae doloribus animi",
          },
          {
            "id": "3e7e8310-4de3-4907-9c7b-806b4d62ccc4",
            "title": "quae vel alias",
          },
        ],
        "error": null,
        "status": "ok",
      },
    },
  },
}
`;

exports[`getServerSideProps without fallback data 1`] = `
{
  "props": {
    "_nextI18Next": {
      "initialI18nStore": {
        "en": {
          "common": {
            "description": "NextJS common setup",
            "no": "No",
            "title": "NextJS Boilerplate | EN",
            "yes": "Yes",
          },
        },
      },
      "initialLocale": "en",
      "ns": [
        "common",
      ],
      "userConfig": {
        "default": {
          "i18n": {
            "defaultLocale": "default",
            "localeDetection": false,
            "locales": [
              "default",
              "en",
              "id",
            ],
          },
        },
        "i18n": {
          "defaultLocale": "default",
          "localeDetection": false,
          "locales": [
            "default",
            "en",
            "id",
          ],
        },
      },
    },
    "fallback": {},
  },
}
`;

exports[`getServerSideProps without locale 1`] = `
{
  "props": {
    "_nextI18Next": {
      "initialI18nStore": {
        "en": {
          "common": {
            "description": "NextJS common setup",
            "no": "No",
            "title": "NextJS Boilerplate | EN",
            "yes": "Yes",
          },
        },
      },
      "initialLocale": "en",
      "ns": [
        "common",
      ],
      "userConfig": {
        "default": {
          "i18n": {
            "defaultLocale": "default",
            "localeDetection": false,
            "locales": [
              "default",
              "en",
              "id",
            ],
          },
        },
        "i18n": {
          "defaultLocale": "default",
          "localeDetection": false,
          "locales": [
            "default",
            "en",
            "id",
          ],
        },
      },
    },
    "fallback": {
      "@"http://localhost:3080/tasks",": {
        "data": [
          {
            "id": "63b8bb76-cf6a-4972-a65a-b5730b69ba27",
            "title": "quidem expedita labore",
          },
          {
            "id": "b86a4d7b-b95b-465a-9533-4fac1c606272",
            "title": "dolores veritatis voluptatum",
          },
          {
            "id": "e6cebf88-7991-4457-aade-4177f3869c17",
            "title": "vel inventore quaerat",
          },
          {
            "id": "48650456-bade-4487-a49d-94a8d215ce32",
            "title": "assumenda quod unde",
          },
          {
            "id": "56034668-b8f0-45ac-890b-e246c355d80a",
            "title": "fugiat quibusdam expedita",
          },
          {
            "id": "2e57a3d4-8dd8-46a5-85d2-6d0578899840",
            "title": "quos beatae adipisci",
          },
          {
            "id": "e5384784-e4f9-44b9-8c56-abd229fd5fd1",
            "title": "repudiandae doloribus animi",
          },
          {
            "id": "3e7e8310-4de3-4907-9c7b-806b4d62ccc4",
            "title": "quae vel alias",
          },
        ],
        "error": null,
        "status": "ok",
      },
    },
  },
}
`;
